libnetwork/ipams/defaultipam/address_space.go

```diff
@@ -2,29 +2,54 @@ package defaultipam
 
 import (
 	"context"
-	"fmt"
 	"net/netip"
+	"slices"
 	"sync"
 
 	"github.com/containerd/log"
 	"github.com/docker/docker/libnetwork/internal/netiputil"
 	"github.com/docker/docker/libnetwork/ipamapi"
+	"github.com/docker/docker/libnetwork/ipamutils"
+	"github.com/docker/docker/libnetwork/ipbits"
 	"github.com/docker/docker/libnetwork/types"
 )
 
 // addrSpace contains the pool configurations for the address space
 type addrSpace struct {
-	// Master subnet pools, indexed by the value's stringified PoolData.Pool field.
+	// Ordered list of allocated subnets. This field is used for subnet
+	// allocations.
+	allocated []netip.Prefix
+	// Allocated subnets, indexed by their prefix. Values track address
+	// allocations.
 	subnets map[netip.Prefix]*PoolData
 
-	// Predefined pool for the address space
-	predefined           []netip.Prefix
-	predefinedStartIndex int
+	// predefined pools for the address space
+	predefined []*ipamutils.NetworkToSplit
 
 	mu sync.Mutex
 }
 
-func newAddrSpace(predefined []netip.Prefix) (*addrSpace, error) {
+func newAddrSpace(predefined []*ipamutils.NetworkToSplit) (*addrSpace, error) {
+	slices.SortFunc(predefined, func(a, b *ipamutils.NetworkToSplit) int {
+		return netiputil.Compare(a.Base, b.Base)
+	})
+
+	// We need to discard longer overlapping prefixes (sorted after the shorter
+	// one), otherwise the dynamic allocator might consider a predefined
+	// network is fully overlapped, go to the next one, which is a subnet of
+	// the previous, and allocate from it.
+	var last *ipamutils.NetworkToSplit
+	var discarded int
+	for i, imax := 0, len(predefined); i < imax; i++ {
+		p := predefined[i-discarded]
+		if last != nil && last.Overlaps(p.Base) {
+			predefined = slices.Delete(predefined, i-discarded, i-discarded+1)
+			discarded++
+			continue
+		}
+		last = p
+	}
```

Comment:
Since the `slices` package is already being used, may as well take full advantage.
```suggestion
	predefined = slices.CompactFunc(predefined, func(last, p *ipamutils.NetworkToSplit) bool {
		return last.Overlaps(p.Base)
	})
```

Reply:
`slices.CompactFunc` works a bit differently. It expects a strict equality as it doesn't compare to the last non-duplicate found, but to 'current-1'. If you have the following subnets:

1. `10.0.0.0/8`
2. `10.0.0.0/16`
3. `10.10.0.0/16`

It tries to compare s1 == s2, and then s2 == s3. That's not what we want.

Reply:
That's too bad. I don't like that `slices.Delete` is being called in a loop:
> Delete is O(len(s)-i)

Therefore the worst case time complexity is O(N^2).

[It is doable in O(len(predefined)) time.](https://go.dev/wiki/SliceTricks#in-place-deduplicate-comparable)

Reply:
Yeah, at first I thought `predefined` wouldn't be filled with that many entries to really matter. But better safe than sorry, I guess. That's now fixed.